### PR TITLE
Added top role property to member

### PR DESCRIPTION
**Pull request by ahnaf-zamil**
### Checklist

- [ ] I have run `nox` and all the pipelines have passed.
- [ ] I have made unittests according to the code I have added/modified/deleted.

### Related issues
This doesn't work stateless without cache. This is why the nox tests couldn't be passed. Also I don't know how to write tests. So I need someone to check the code and see if it has issues.

### CODE REVIEW

File: hikari/guilds.py [hunk 1]
```diff
@@ -421,6 +421,21 @@ def mention(self) -> str:
...
+    def top_role(self) -> Role:
+        """Returns the top role of the given member.
+
+        Returns
+        -------
+        hikari.guilds.Role
+            The highest role the member has.
+        """
+        roles = [r for r in self.app.cache.get_roles_view_for_guild(self.guild_id).values() if r.id in self.role_ids]
+        return max(
```
Comment (davfsa) [inline hikari/guilds.py:434]: Please run black `nox -s black` to fix this as the linter is complaning

File: tests/hikari/test_guilds.py [hunk 2]
```diff
@@ -161,6 +162,9 @@ def test_mention_property_when_no_nickname(self, obj):
         obj.nickname = None
         assert obj.mention == "<@123>"
 
+    def test_top_role(self, obj):
+        assert obj.top_role.guild_id == "Owner"
```
Comment (davfsa) [inline tests/hikari/test_guilds.py:166]: This is not the right way of testing this. You will have to mock `app.cache.get_roles_view_for_guild`, make it return a dict with the values you expect, make sure it gets called with the right guild_id and that the role return is the correct one. You will also have to write tests (and modify the code you wrote) for when an empty role view is returned (missing intents so we didnt cache it or he is using a stateless bot).

**Review by davfsa — CHANGES_REQUESTED** [review]
Several things and you will have to make it take into account a missing intents cache and a stateless bot (which will return a EmptyCacheView) and return `None` instead of a role. Also, always remember to run `nox` before you commit to make sure all pipelines will pass

**Comment by nekokatt** [comment]
You are getting the hang of how these components slot together, just a couple of points that I'd like to see fixed from me and Dav, if that is okay?

Thanks for spending the time to do this, it is greatly appreciated :-)

**Review by nekokatt — CHANGES_REQUESTED** [review]

**Comment by ahnaf-zamil** [comment]
> 
> 
> You are getting the hang of how these components slot together, just a couple of points that I'd like to see fixed from me and Dav, if that is okay?
> 
> Thanks for spending the time to do this, it is greatly appreciated :-)

Thanks, I just tried to do something to help. I am not very familiar with caching and testing in general so it's kinda hard for me to contribute to this project.

**Comment by nekokatt** [comment]
> > You are getting the hang of how these components slot together, just a couple of points that I'd like to see fixed from me and Dav, if that is okay?
> > Thanks for spending the time to do this, it is greatly appreciated :-)
> 
> Thanks, I just tried to do something to help. I am not very familiar with caching and testing in general so it's kinda hard for me to contribute to this project.

if you want any help, ping one of us :+1:

File: tests/hikari/test_guilds.py [hunk 3]
```diff
@@ -105,6 +105,7 @@ def obj(self):
                 flags="flags",
                 mention="<@123>",
                 __str__=mock.Mock(return_value="davfsa#0001"),
+                top_role=mock.Mock(return_value="Owner"),
```
Comment (davfsa) [inline tests/hikari/test_guilds.py:108]: Dont mock something you are testing ;)
